heroic-component/src/main/java/com/spotify/heroic/metric/QueryResult.java

```diff
@@ -95,8 +114,51 @@
                 limits.add(ResultLimit.GROUP);
             }
 
-            return new QueryResult(range, groupLimit.limitList(groups), errors, trace,
-                new ResultLimits(limits.build()));
+            QueryResult queryResult = new QueryResult(range, groupLimit.limitList(groups), errors,
+                                                      trace, new ResultLimits(limits.build()));
+
+            queryResult.logQueryInfo(request, requestMetadata,
+                                     logQueries, logQueriesThresholdDataPoints);
+
+            return queryResult;
         };
     }
+
+    public void logQueryInfo(final FullQuery.Request request,
+                             final QueryRequestMetadata requestMetadata,
+                             final boolean logQueries,
+                             final OptionalLimit logQueriesThresholdDataPoints
+    ) {
+        if (!logQueries) {
+            return;
+        }
+
+        totalQueriesProcessed.increment();
+
+        long totDataPoints = 0;
+        for (ShardedResultGroup g : groups) {
+            totDataPoints += g.getMetrics().getData().size();
+        }
+
+        if (totDataPoints < logQueriesThresholdDataPoints.orElse(
+                                    OptionalLimit.of(0)).asLong().get()) {
+            return;
+        }
+
+        long currQueriesAboveThreshold = queriesAboveThreshold.incrementAndGet();
+
+        boolean isIPv6 = requestMetadata.getRemoteAddr().indexOf(':') != -1;
+        queryLog.debug("[Query] (" + currQueriesAboveThreshold + "/" + totalQueriesProcessed + ")" +
```

Comment:
Also, use `trace` log level.

Don't align statements like this. Use the format generated by the provided [code-style](https://github.com/spotify/heroic/tree/master/idea) or escape it using `@formatter:off` comments as described in the [README](https://github.com/spotify/heroic#bypassing-validation).

Special formatting (without escaping) causes future issues when someone tries to build on top of your change. And they accidentally format your code, making their change harder to decipher.